### ISSUE TITLE
Allow logging of the number of requests a connection has seen

### ISSUE THREAD

**Issue by ross** [haproxy/haproxy#1050]
## What should haproxy do differently? Which functionality do you think we should add?

It would be useful when tuning settings related to keep alive/timeouts if there was a log format available in HAProxy with the request number per connection. E.g. `connection_request_count=%cr` or whatever variable name would make the most sense. The variable would be 1 for the first request across the connection, 2 for the second, ... 

It'd be somewhat similar to "%rt" and "%lc" except that it would be per connection and not hexadecimal and not counting the number of requests handled by the process (I'm assuming here the docs aren't super clear on those.)

## What are you trying to do?

Get a better view of how how many requests are send down a single connection. It'd be useful both for tuning keep-alive timeouts and for generally debugging things that involve using a single connection for multiple requests. 

The best we've been able to do absent that is looking at ~proxies like %Tt much larger than %Ta, but other things can skew those attempts. 

## Output of `haproxy -vv` and `uname -a`

```
$ /usr/sbin/haproxy -vv
HA-Proxy version 2.0.19+foo2 2020/11/06 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -D_FORTIFY_SOURCE=2 -fstack-protector-strong -fPIC -DMAX_SYSLOG_LEN=2048
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_THREAD=1 USE_PTHREAD_PSHARED=1 USE_REGPARM=1 USE_STATIC_PCRE2=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_ZLIB=1 USE_TFO=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD +PTHREAD_PSHARED +REGPARM -STATIC_PCRE +STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL -LUA +FUTEX +ACCEPT4 -CLOSEFROM -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=32).
Built with OpenSSL version : OpenSSL 1.1.1i  8 Dec 2020
Running on OpenSSL version : OpenSSL 1.1.1i  8 Dec 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE        mux=H2
              h2 : mode=HTX        side=FE|BE     mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services : none

Available filters :
        [SPOE] spoe
        [COMP] compression
        [CACHE] cache
        [TRACE] trace
$ uname -a
Linux glb-proxy-6092504.va3-iad.github.net 4.19.0-0.bpo.10-amd64 #1 SMP Debian 4.19.132-1~deb9u2 (2020-08-11) x86_64 GNU/Linux
```

**Comment by wtarreau** [comment]
I agree and have been missing this feature as well. However when I was about to add it, I figured that it's not that clearly defined anymore now that we support HTTP/2. Indeed, requests happen in parallel for a given connection, and there is already a reliable identifier there, which is the stream ID. This number usually increments in two (odd numbers only are used for streams initiated by the client).

If we'd use the connection's request counter instead, the counter would happen in any order and with jumps. For example if we receive 5 requests, the counter goes to 5, then they're processed then logged, all of them would present "5" in the logs. The only guarantee it would give would be that the last logged request would present the largest known value.

But what I understand is that you essentially need to easily see in your logs how many different requests were sent over a given connection, so you probably don't care about the exact number seen there as long as it's easy enough to check and count, which the stream ID already matches better than the request counter. We could then implement a simple counter for HTTP/1 to provide the same feature. Several of us have already been discussing a high-level interface to retrieve such information for logging purposes.

**Comment by ross** [comment]
> But what I understand is that you essentially need to easily see in your logs how many different requests were sent over a given connection, so you probably don't care about the exact number seen there as long as it's easy enough to check and count, which the stream ID already matches better than the request counter. We could then implement a simple counter for HTTP/1 to provide the same feature. Several of us have already been discussing a high-level interface to retrieve such information for logging purposes.

Yeah. A combination of stream ID (for HTTP/2) and an incrementing counter (for HTTP/1) sounds like it'd suffice.

**Comment by capflam** [comment]
FYI, It was implemented. It is available in 2.9 and 3.x. `fs.id` and `bs.id` should be used to retrieve it.  The stream ID is returned for H2/H3/FCGI connections and the number or processed requests for H1 connections.
